### PR TITLE
Add clustering scripts to tools

### PR DESCRIPTION
**Pull request by ArnauBigas**
This PR adds the python scripts for doing the clustering from my repository (https://github.com/ArnauBigas/chopstix-trace-inspect) to this one so that everything is in the same place and is easier to manage.

As discussed I have added them to the tools folder, within a subdirectory called "clustering". This is just a 1 to 1 copy, I am open to changing any filename or the structure in any way :)

Additionally, this should probably be integrated with the CMake scripts so that it is installed properly, but I have never used python in combination with CMake.

### CODE REVIEW

**Comment by rbertran** [comment]
Thanks Arnau, I'll work on it on this PR in the following weeks.

**Comment by rbertran** [comment]
@ArnauBigas I'm trying to integrate these scripts and test them on the IBM platforms (POWER and Z). I do have some questions and some TODOs that I think we can be implementing together in our endless free time ;)

I'll use this branch as a WIP, once all the features below are in place, we can integrate it in the main repo. So, far, all the scripts work great except some minor usability and automation tips (and cross-platform usability). Let me know on which features you are already working or you would like to work so that we don't step on each other's work. I ordered the list by priority.

- [ ] the tool `riscv-perf` is referenced in the documentation, but I don't know where I can find it. Is in another branch/repo? I assume that we need to develop a similar one for P and Z, is that correct?  
- [ ] function `estimate_dbscan_epsilon` triggers errors. I think it assumes invocation count but the matrix has invocation "sets". Also, the coverage computation, at end of the function, trigger an out-of-bounds error as well. Maybe is related to the aforementioned issue.
- [ ] Provide a high level script that automates the whole process. Now that we have clustering in place, we can implement a generic script that given a command does the following:
    - trace the command and gets the top functions
    - trace the invocations of each of the top functions 
    - selects the representative invocations for each function
    - trace the selected invocations
    - generates mpts for the selected invocation
    - generates elfs for the selected invocations
    - runs/validates elfs of the selected invocation
    - generates the final report about all the stats (per benchmark, per function, etc.)
- [ ] Provide a step-by-step quide on how you'd use these tools (the path from how to generate a trace to the final step to select the few representative invocations)
- [ ] Integrate the above documentation with the main one. 
- [ ] command line documentation can be improved by providing not only the high level description of the commands and options but also what is the intended use.
- [ ] for the couple of tools that generate charts, it'd good to add options or guidance in the documentation on how to save them to other format such as pdf/eps/jpeg and modify title/sizes/etc.

**Comment by ArnauBigas** [comment]
I've taken a look through the changes, they're all very much welcome ones. Thanks a lot!

I'll answer the questions in more or less the same order:

- I changed the name of the tool, since I had to port it to work on other architectures for using these tools on POWER9, but apparently I forgot to change the documentation. The name used in the documentation should be changed to `perf-invok`. You can find the source code for this tool in https://github.com/ArnauBigas/perf-invok. I don't mind integrating that source code with this repo (originally the clustering scripts were also in another repo).
- The function you're mentioning was the first one developed, quite a while ago. It didn't work very well so it got abandoned. Afterwards, the "Invocation Set" concept was introduced, and this function wasn't updated. I don't use it with any script, so we can either remove it or fix it.
- I have recently developed an internal tool that generates a script of sorts like you mention (used for submitting batches and such, lots of benchmark and platform-specific pieces are assembled together into a bash script). I could generate a simple one and use that as a base for what you're suggesting. Aside from that, all of the points/steps make sense to me.
- The rest of the documentation points are very valid, I suppose I'll work on them at some point.
- The tools for generating the charts are pretty much "quick" hacks without documentation. I agree that they should have some documentation, at least answering:
  - What files are used as input
  - What is the expected file structure
  - What is the output
  - Not part of the documentation, but it'd be good to add options to set the format the charts will be saved as (if it's trivial).

**Comment by rbertran** [comment]
Thank for the quick answer. The changes committed are _minor_ , mostly related on format, headers and some incompatibilities I found in Z. Hope they don't break anything on your side ;)

- I personally prefer the `perf-invok` to sit within **ChopStix** world as it seems to be tightly coupled with the its functionality and goals. That said, I can see use cases where someone wants to profile performance counters at code-level (from custom start to end address) and in such case, this tool comes pretty handy, and does not require all the `chopstix` extras. Since you are the project founder and the BDFL ;), you make the call, and I'll integrate stuff accordingly. 
- btw, I've not had time to use test `perf-invok` ... I assume that the output format is compatible with the clustering `.py` scripts of this PR. 
- So, `estimate_dbscan_epsilon` is actually _deprecated_ according to your experience so far. So, I'd remove it and simplify scripts/functionality (e.g. no need for `coverage` parameter). Is that OK? At the same time, we do need to provide a default parameter for `epsilon` and document (in the command line as well as in the doc) what values are used typically. (e.g. 0.1, 0.01, 0.001, 0.0001, etc.) This way the user has some guidance of what `epsilon` should be. Which values do you typically use for epsilon? 
- I've developed a [script](https://github.com/rbertran/hacl-star/blob/chopstix/tests/trace_tests.sh) too but it is not generic at all, I did it quite fast and _dirty_ ... I think your approach of having a _driver_ (main script) that generates a script that then can be submitted scales better and seems more flexible (you can choose the level of granularity too). Overall, the best would be to implement a new script from scratch with proper checking of results/error/steps and some query support to check if something has already been `chopstix`-ed . Please share your script (email or here) and I can take a look, to see if something can be reused. My idea is to rely on a _predefined_ directory structure that contains all results, steps etc.  
- Don't worry about the charting tools, they are perfect for now, we'll _tune_ to generate nice reports later on. 

Thanks a lot!!!

File: tools/clustering/src/clustering.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+                count = len(dp[0])
+                dp[0].append(invocation)
+                dp[1] = ((dp[1] * count) + invocation.metrics.instructions) / (count + 1)
+                dp[2] = ((dp[2] * count) + invocation.metrics.ipc) / (count + 1)
+                dp[3].append(i)
+                done = True
+        if done == False:
+            data.append([[invocation], invocation.metrics.instructions, invocation.metrics.ipc, [i]])
+
+    X = np.array([[dp[2] * 20, math.log(dp[1])] for dp in data])
```
Comment (rbertran) [inline tools/clustering/src/clustering.py:219]: @ArnauBigas Here, I assume that you try to normalize the range of both metrics: ipc and # of instructions. What is the reason to use 20 and log? Is this something you _tunned_ experimentally, if so, we need to expose such "normalization" options to the user. Maybe other experiments, require other _tunning_ ...

That said, I'm of the opinion of using ageneric approach. Can we use data normalization and/or standardization on so that both metrics have the same _weight_? [1] . Maybe we already discussed this in the past and the results were not satisfactory(?). In any case, if that is considered, we need to remove outliers on the data before doing the transformation (and assign such data points directly as noise), otherwise the clustering may be affected. 

[1] https://python-data-science.readthedocs.io/en/latest/normalisation.html

File: tools/clustering/src/clustering.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        for dp in data:
+            if not done and abs(1 - (invocation.metrics.instructions / dp[1])) < 0.01:
+                count = len(dp[0])
+                dp[0].append(invocation)
+                dp[1] = ((dp[1] * count) + invocation.metrics.instructions) / (count + 1)
+                dp[2] = ((dp[2] * count) + invocation.metrics.ipc) / (count + 1)
+                dp[3].append(i)
+                done = True
+        if done == False:
+            data.append([[invocation], invocation.metrics.instructions, invocation.metrics.ipc, [i]])
```
Comment (rbertran) [inline tools/clustering/src/clustering.py:217]: @ArnauBigas Please explain this peace of code and why is performed. If I understand correctly, seems to me that you _pre-cluster_ the invocations by instruction count?  is this done to speedup the following clustering step?

File: tools/clustering/src/clustering.py (hunk 1 above)
Comment (ArnauBigas) [reply tools/clustering/src/clustering.py:219]: You are mostly right. It was tuned experimentally, based on the intuition that IPC would have a range of (0, 4?), while the retired instruction count would be in the millions or even billions. Thus, the reason for the log() and the smaller adjustment on the IPC.

After having had some more university courses on this subject since I wrote this, maybe we could try using a standard scaler such as [minmax scaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html)? We would have to experimentally confirm it works, though.

File: tools/clustering/src/clustering.py (hunk 2 above)
Comment (ArnauBigas) [reply tools/clustering/src/clustering.py:217]: That is correct, sklearn can't easily deal with clustering some of the millions of invocations found in some benchmarks. Also, keep in mind that some of the machines we use in our use-case are relatively weak ;)

So, yeah, these are optimizations to speed-up the clustering.

File: tools/clustering/src/clustering.py (hunk 1 above)
Comment (rbertran) [reply tools/clustering/src/clustering.py:219]: Yupp, scaling the range to [0,1] (after removing outliers) will be a more _generic_ solution since it will give both metrics (ipc and # instructions) the same _weight_. We can try the minmax scaler, but it may generate problems by changing the sparsity of the data. For example, in the case where we have only two big values (e.g. `[400000010 ,400000000]`), the transformation will be `[0,1]` and that will _misslead_ the clustering algorithm (the points are actually very close). I'm more inclined to use the  [MaxAbsScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MaxAbsScaler.html#sklearn.preprocessing.MaxAbsScaler), since it does not destroy any sparsity. Obviously, we need to try it experimentally and see if this is the case...

Comment (ArnauBigas) [reply tools/clustering/src/clustering.py:219]: That's a very good point, that scaler might work better.

However, regardless of whatever scaler we use, I'm worried about the "epsilon" parameter of the clustering algorithm (DBSCAN). This parameter "defines the minimum distance between two points of different clusters" (intuitively describing what I know the parameter effectively does). The parameter right now is user-defined, and hard-coded in my scripts. I'm not sure if we should try to find the optimum value for this parameter instead? This would induce a high overhead, though, and the hardcoded value works quite fine so far with the "experimentally tuned scaler"...

File: tools/clustering/src/clustering.py (hunk 2 above)
Comment (rbertran) [reply tools/clustering/src/clustering.py:217]: Now, I get it. Note that the optimization _prioritizes_  the #instructions metric since it _averages_ and _joins_ invocations with similar #instructions that might have quite different IPC. This is not a problem as we have seen experimentally. In fact, I guess that that rarely happens (having the same #instructions and significantly different IPC) .  Intuition tells me that the same #instructions indicate that the same code path has been executed, and therefore we should expect the same performance.
That makes me wonder, have we ever tried to use #instructions alone as a clustering approach similar to what we did for IPC alone? If that turns out to be as good as this approach, it will be better since the #instruction metric is much less prone to measurement/experimental noise.

Comment (ArnauBigas) [reply tools/clustering/src/clustering.py:217]: Yes, you understood that correctly. The argument for prioritizing the #instructions is the one I used when I wrote that bit of code. Similar amount of instructions has a high probability of being the same execution path, although coincidentally two different code paths could have the same #instructions.

These are the clustering ideas we have used so far:
1. Clustering based on #page accesses and the address of each access.
2. Clustering based on IPC (prone to the noise you mention)
3. Clustering based on IPC & #instructions (the one we're using right now)

So, no, we haven't used instructions alone ;)

File: tools/clustering/src/clustering.py (hunk 1 above)
Comment (rbertran) [reply tools/clustering/src/clustering.py:219]: I see, you already tuned the epsilon for the experiments done so far and we do not want to redo all that. So, we keep this clustering algorithm as is. We can implement modifications/improvements in a different clustering strategy (i.e. in a different function). 

Regarding about finding the optimum epsilon automatically, I'm totally in. My goal is to have a good enough _push button_ solution without requiring any use input/expertise (just the command ;) ) So, I'm thinking in something along the following lines: Perform a _greedy_ search of different eps at the expense of execution time. Start with a _strict_ eps (many clusters or many noise points), and iteratively modify eps until: (a) no more than X clusters are defined and (b) no more than %X of points are noise. Does that makes sense?

File: tools/clustering/cti-cluster-info.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        for invocation in cluster_info.get_random_noise_invocations():
+            print(invocation)
+    elif args.cluster != None:
+        print(cluster_info.get_random_invocation_in_cluster(args.cluster))
+    else:
+        for index in range(cluster_info.get_cluster_count()):
+            print(cluster_info.get_random_invocation_in_cluster(index))
+
+        for invocation in cluster_info.get_random_noise_invocations():
+            print(invocation)
```
Comment (rbertran) [inline tools/clustering/cti-cluster-info.py:50]: Right know we are returning a random representative, which is OK since it is expected that all elements in a cluster behave similarly. However, there are to drawbacks:
- It complicates validation (2 successive runs on the same data will generate different outputs)
- It might introduce error when for _sparse_ clusters. Let's say that for whatever reason, the clustering algorithm comes up with a cluster definition where the data points within the cluster are quite sparse (note that they are not for the entire data set and that's why they are in the same cluster)

I propose to return always the _center_ of the cluster. We can add a flag to select the behavior we want.

**Comment by rbertran** [comment]
To summarize the latest commits on this branch:

- Added plotting capability in all clustering algorithms except "trace"
- A bit of code cleanup
- Added two new clustering approaches on top of the existing 3 approaches (trace, DBSCAN IPC, DBSCAN IPC&INSTR):
  - DBSCAN on # instructions that we discussed earlier with some "auto-tunning" of the EPS parameter and data normalization
  - Density map using the 2D dimensions: IPC and # instructions, and weighting clusters by # Instructions (prioritize invocations with more instructions). Also, outliers are discarded. 
 - Added a new function to compute extra results on the JSON file:
   - coverage: compute % of invocations within a cluster
   - coverage: compute % of #instruction covered within a cluster
   - for each cluster, the centroid (according to the metrics used) is computed
   - the noise_invocations are sorted by the # instructions. We can still pick one randomly, or based on the # instructions.
   - the invocations set for each cluster are sorted by the distend to the centroid to the cluster. We can still pick one randomly, or based on this distance. 
 - Input CSV files can be gzipped
 - Formatted all python files automatically using `black` formatter
 - Improved the CLI, added help, etc.
 - rename cti-dbscan.py to cti-cluster.py, scine we do not only use DBSCAN
 
 I've not touched existing algorithms, so I hope these changes do not generated a major break in your code.
